pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml

```diff
@@ -0,0 +1,343 @@
+            i--;
+            ++i;
+            --i;
+            i += 1;
+            i -= 1;
+            i *= 1; // not OK
+            i /= 1; // not OK
+            i = 5; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: reassigned 'foreach' loop variable</description>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>6</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            doSomethingWith(i);
+            i = 5; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: increment 'foreach' loop variable, but not as first statement</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>6</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            doSomethingWith(i);
+            i += 1; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>no violation: increment 'foreach' loop variable as first statement</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            i++;
+            doSomethingWith(i);
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>no violation: assign-increment 'foreach' loop variable as first statement</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            i += 1;
+            doSomethingWith(i);
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: increment 'foreach' loop variable multiple times</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>7</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            i++;
+            doSomethingWith(i);
+            i++; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: pre-increment 'foreach' loop variable multiple times</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
```

Comment:
I'm not sure firstOnly is really useful for foreach loops. My 2 cents: 
* AFAIK it's uncommon to use a foreach loop to iterate over an array of numbers anyway
* The  `i` in foreach loops is not an index variable, but an item of an iterable. In any case I can think of, it's confusing and not useful to reassign it:
   * it doesn't change the value in the iterable
   * the same thing can be achieved by `doSomethingWith(i + 5)`, without reassignment.
   * if really needed, you can always declare another variable inside the loop `int shifted = i + 5`, which clarifies the code and preserves the original item

In fact, I'm interested in the behaviour of the rule with other data types, e.g.
```java
    void foo(List<Node> bar) {
        for (Node n : bar) {
            n = n.jjtGetParent(); // I can't think of a coding standard where this wouldn't be considered bad practice
            // better do Node parent = n.jjtGetParent();
            doSomethingWith(n);
        }
    }
```

Do you really see a use case for configurability of the reassignment policy in foreach loops, besides `allow`/`deny`?

Reply:
OK, iterating over numbers really isn't a good example.

What I want to achieve with having `firstOnly` is the ability to write
```java
for (String stringWithSpaces : listOfStrings()) {
    final String string = stringWithSpaces.trim();
    // keep working with string
}
```
as
```java
for (String string : listOfStrings()) {
    string = string.trim();
    // keep working with string
}
```

In the first example, `stringWithSpaces` still exists and could be used (even if it shouldn't). In the second example the code will always use the cleaned value.

The `string.trim()` is only representative for some kind of transformation/clean-up and only useful if the variable is used more than once.
So this version
```java
    void foo(List<Node> bar) {
        for (Node n : bar) {
            doSomethingWith(n.jjtGetParent());
        }
    }
```
is preferable to the reassignment of the loop variable.

It is an edge-case, but a valid one, I think.

Reply:
You've convinced me. I originally thought keeping the two variables in scope was an advantage, but you're right that for long loops it's better to reduce the number of variables in scope. Thanks for explaining